File: getting-started-mocks/01-first-useful-mock/node/notifier/notifier.js
// Turns a shipment status into a message worth sending.
//
// `getShipmentStatus` is the seam: an async function that answers "where is
// this package?" for one tracking number. Production wires it to the carrier's
// HTTP API (see carrier.js). A test hands in whatever answer it needs.
export class Notifier {
  #getShipmentStatus

  constructor(getShipmentStatus) {
    this.#getShipmentStatus = getShipmentStatus
  }

  // Returns the message to send, or null when nothing needs saying.
  async notify(trackingNumber) {
    const status = await this.#getShipmentStatus(trackingNumber)
    if (status === 'delayed') {
      return `Package ${trackingNumber} is delayed`
    }
    return null
  }
}

File: getting-started-mocks/01-first-useful-mock/python/notifier/notifier.py
"""Turns a shipment status into a message worth sending."""

from collections.abc import Callable

# Answers "where is this package?" for one tracking number.
#
# This is the seam. Production wires it to the carrier's HTTP API (see
# carrier.py). A test hands in whatever answer it needs.
ShipmentStatus = Callable[[str], str]


class Notifier:
    def __init__(self, get_shipment_status: ShipmentStatus) -> None:
        self._get_shipment_status = get_shipment_status

    def notify(self, tracking_number: str) -> str | None:
        """Return the message to send, or None when nothing needs saying."""
        if self._get_shipment_status(tracking_number) == "delayed":
            return f"Package {tracking_number} is delayed"
        return None
